### PR TITLE
simplify core

### PR DESCRIPTION
**Pull request by rowanG077**
- [x] Rewritten core to handle metadata and key generation separately from streaming
- [x] Added tests
- [x] Rewrite core tests
- [x] Fix dependent packages

### CODE REVIEW

File: irmaseal-core/src/metadata_reader.rs [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+                unconsumed: buf_size - consumed,
+                header: core::mem::take(&mut self.header_buf),
+                metadata: m,
+            })
+        } else {
+            Ok(MetadataReaderResult::Hungry)
+        }
+    }
+
+    fn parse_meta_data(&self) -> Result<Metadata, Error> {
```
Comment (leonbotros) [inline irmaseal-core/src/metadata_reader.rs:87]: style: `parse_meta_data` -> `parse_metadata`

**Review by leonbotros — CHANGES_REQUESTED** [review]
Looks good! Some minor changes requested.

File: irmaseal-core/Cargo.toml [hunk 2]
```diff
@@ -29,4 +28,4 @@ ibe = "0.1.3"
 [features]
 default = ["stream"]
 
-stream = ["digest", "rand", "aes", "ctr", "hmac", "sha3"]
+stream = ["aes", "ctr", "cfb-mode" ]
```
[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-core/Cargo.toml:31]: Yes I think. I will check this

File: irmaseal-cli/src/decrypt.rs [hunk 3]
```diff
@@ -67,14 +82,18 @@ pub async fn exec(m: &ArgMatches<'_>) {
...
-            crate::util::FileUnsealerRead::new(sealed.unseal(&metadata, &r.key.unwrap()).unwrap());
+        let mut unsealer = crate::util::FileUnsealerRead::new(
+            Unsealer::new(&metadata, header, &kr.key.unwrap(), r).unwrap(),
+        );
 
-        let mut tar = Archive::new(unsealed);
-        tar.unpack(".").unwrap();
+        std::io::copy(
+            &mut unsealer,
+            &mut std::fs::File::create("./unsealed").unwrap(),
```
[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-cli/src/decrypt.rs:94]: Trye

File: irmaseal-core/src/lib.rs [hunk 4]
```diff
@@ -13,41 +14,61 @@ pub mod stream;
...
+
+type HeaderBuf = ArrayVec<[u8; MAX_HEADERBUF_SIZE]>;
+
+// How large a block is that has to be encrypted using
+// the symmetric crypto algorithm
+pub(crate) const SYMMETRIC_CRYPTO_BLOCKSIZE: usize = 16384;
+
+// Which symmetric crypto algorithm is used.
+#[allow(dead_code)]
+pub(crate) const SYMMETRIC_CRYPTO_IDENTIFIER: &str = "AES256-CTR";
```
[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-core/src/lib.rs:70]: `AES256-CTR64BE`?

[the remark this answers is not in the document]
Comment (leonbotros) [reply irmaseal-core/src/lib.rs:70]: I agree, that would be most compact and it includes everything necessary

[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-core/src/lib.rs:70]: It does not include how to initialize the counter but maybe that is too specific?

[the remark this answers is not in the document]
Comment (leonbotros) [reply irmaseal-core/src/lib.rs:70]: If you include `CTR64-BE`, any consumer library/application can initialize the counter if they want to. If some application wants to start at 0 and perform additions that are checked for overflow, one could pass an IV with the last 8 bytes set to the representation of `0u64.to_be_bytes()` for example.

[the remark this answers is not in the document]
Comment (leonbotros) [reply irmaseal-core/src/lib.rs:70]: I.e., by ignoring some of the IV from a `metadata`.

**Review by leonbotros — APPROVED** [review]

**Pull request merged**